### PR TITLE
feat(ssh): public key check ( #509 )

### PR DESCRIPTION
**Pull request by faust64**
TBH: did not test it yet. TODO....

### CODE REVIEW

File: lib/functions.inc.php [hunk 1]
```diff
@@ -519,6 +519,38 @@ function change_password( $ldap, $dn, $password, $ad_mode, $ad_options, $samba_m
...
+    if (! in_array($algorithm, array('ssh-rsa', 'ssh-dss'))) {
+	return false;
+    }
+
+    $key_base64_decoded = base64_decode($key, true);
+    if ($key_base64_decoded == FALSE) {
+	return false;
+    }
+
+    $check = base64_decode(substr($key,0,16));
```
Comment (plewin) [inline lib/functions.inc.php:545]: this will break for anything that is not ssh-rsa or ssh-dss

File: htdocs/changesshkey.php [hunk 2]
```diff
@@ -40,8 +40,10 @@
 }
 if (isset($_POST["password"]) and $_POST["password"]) { $password = strval($_POST["password"]); }
  else { $result = "passwordrequired"; }
-if (isset($_POST["sshkey"]) and $_POST["sshkey"]) { $sshkey = strval($_POST["sshkey"]); }
- else { $result = "sshkeyrequired"; }
+if (isset($_POST["sshkey"]) and $_POST["sshkey"]) {
+    $sshkey = strval($_POST["sshkey"]);
+    if (! check_sshkey($sshkey)) { $result = "invalidsshkey"; }
```
Comment (plewin) [inline htdocs/changesshkey.php:45]: I don't feel confident about having a mandatory validation if the validation is not unit tested with a range of public keys.

**Review by plewin — CHANGES_REQUESTED** [review]
I believe check_sshkey should be unit tested before considering a merge because it is a difficult validation to implement
ssh-ed25519 public keys do not currently validate while they should

File: lib/functions.inc.php [hunk 3]
```diff
@@ -519,6 +519,38 @@ function change_password( $ldap, $dn, $password, $ad_mode, $ad_options, $samba_m
...
 
 
+# Verifies sshPublicKey is valid
+function check_sshkey ( $sshkey ) {
+    $key_parts = explode(' ', $sshkey, 3);
+
+    if (count($key_parts) < 2) {
+	return false;
+    }
+    if (count($key_parts) > 3) {
```
[the remark this answers is not in the document]
Comment (faust64) [reply lib/functions.inc.php:529]: Have you tested this?

```
$ cat test.php

<?php
$str1 = "";
$str2 = " ";
$str3 = "   	  "; #with a tab in there
$testkey = "ssh-rsa AAAAB blabla";

$len1 = count(explode(' ', $testkey, 3));
$len2 = count(explode(' ', $str1, 3));
$len3 = count(explode(' ', $str2, 3));
$len4 = count(explode(' ', $str3, 3));

echo "good=$len1\n";
echo "bad1=$len2\n";
echo "bad2=$len3\n";
echo "bad3=$len4\n";
?>

$ php test.php 
good=3
bad1=1
bad2=2
bad3=3
```

Now, we can remove the limit in the explode.

File: lib/functions.inc.php [hunk 4]
```diff
@@ -519,6 +519,38 @@ function change_password( $ldap, $dn, $password, $ad_mode, $ad_options, $samba_m
...
+	return false;
+    }
+
+    $algorithm = $key_parts[0];
+    $key = $key_parts[1];
+
+    if (! in_array($algorithm, array('ssh-rsa', 'ssh-dss'))) {
```
[the remark this answers is not in the document]
Comment (faust64) [reply lib/functions.inc.php:536]: I'll move valid key types to a variable in config

File: lib/functions.inc.php (hunk 1 above)
Comment (faust64) [reply lib/functions.inc.php:545]: let's pick 27 chars then

File: lib/functions.inc.php (hunk 3 above)
[the remark this answers is not in the document]
Comment (faust64) [reply lib/functions.inc.php:529]: although, if I remove the limit in the explode and that user mistakenly inserts two consecutive spaces, then a valid key would be refuse.
I think I'll keep both explode limit and count checks. Still writing some tests...

[the remark this answers is not in the document]
Comment (faust64) [reply lib/functions.inc.php:529]: Let's go with preg_split instead. Running some tests, this one sounds more reliable...

File: htdocs/changesshkey.php (hunk 2 above)
Comment (faust64) [reply htdocs/changesshkey.php]: CI would run tests for dsa, ecdsa, rsa, ed22519.
Let me know if you have others in mind.

File: tests/CheckSshkeyTest.php [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+    {
+	# Load functions
+	require_once("lib/functions.inc.php");
+
+	# Valid SSH Keys Configuration
+	$ssh_valid_key_types = array('ssh-rsa', 'ssh-dss', 'ecdsa-sha2-nistp256', 'ssh-ed25519');
+
+	# Sample inputs to test
+	$valid_key_dsa = "ssh-dss    AAAAB3NzaC1kc3MAAACBAPk0v+w59ie8OmBDlWdBKq/8YFP9h6BbnDxUkwLNlUQ27sCdI3Ko6R/ivEC22Q+SKSF+vL8TThfRWlwhDuro0aqK/0oqwz0d8wyxwWKSlygzME7PoAA5xIxAwTSF3apxGhMkZ3xVQyVVO0P/w4nRGOYEEfmeZdR6NMtZej9UszmfAAAAFQC8QxCoKLq+0e23E50YpO+gEHKrqwAAAIAGvgIwBXV7SxhhdudjJKBpjSVc+SOM5oxMT/Yx8qfkx6nGrvsmBMtCtjkDYsYxfC/GAJK2LEvfPUah5a0U33EmfIXxSjc+MGOd90i69FKSfaa5nuGb+9T/bhevYDsJvBkvl+rWABLCnDXvqUY0ZJ5JO6mT9xD+mq7Q7hy6kZzVAgAAAIARAuLXJVf3tMsASLxm++efey9dQl+8EsveT08E45eKxJU0F5yIdqkHcuWquDP9xI9e35cKx13W5uSGIr9TDH9wNjyY0aX3MkUeYM9UQut3bHr/tH4TjM87hDitOCuRERGZwvvhHX3mxzug0N44DLyXmZFvV2Wq+MInLdVujcricg== a@b";
+	$valid_key_ecdsa = "ecdsa-sha2-nistp256 AAAAE2VjZHNhLXNoYTItbmlzdHAyNTYAAAAIbmlzdHAyNTYAAABBBCt5Z3IMcaL0OAQ62QpGLvRE+5GsolzqT7gGQ/76gz348IJfb+6MyXPQbT3BmNLtSw88UaaKp0HOQFj6f3US8rU= a@b";
```
Comment (plewin) [inline tests/CheckSshkeyTest.php:18]: Can you please add ecdsa-sha2-nistp384 and ecdsa-sha2-nistp521

File: conf/config.inc.php [hunk 6]
```diff
@@ -168,6 +168,9 @@
 # What attribute should be changed by the changesshkey action?
 $change_sshkey_attribute = "sshPublicKey";
 
+# Ensure the SSH Key submitted uses a type we trust
+$ssh_valid_key_types = array('ssh-rsa', 'ssh-dss', 'ecdsa-sha2-nistp256', 'ssh-ed25519');
```
Comment (plewin) [inline conf/config.inc.php:172]: Can you please add ecdsa-sha2-nistp384 ecdsa-sha2-nistp521

**Review by plewin — CHANGES_REQUESTED** [review]
This MR work. Have a slight UX issue with the message color.
Oddly this PR is highly inconsistent with space and tabs

File: htdocs/changesshkey.php [hunk 7]
```diff
@@ -40,8 +40,10 @@
 }
 if (isset($_POST["password"]) and $_POST["password"]) { $password = strval($_POST["password"]); }
  else { $result = "passwordrequired"; }
-if (isset($_POST["sshkey"]) and $_POST["sshkey"]) { $sshkey = strval($_POST["sshkey"]); }
- else { $result = "sshkeyrequired"; }
+if (isset($_POST["sshkey"]) and $_POST["sshkey"]) {
+    $sshkey = strval($_POST["sshkey"]);
+    if (! check_sshkey($sshkey, $ssh_valid_key_types)) { $result = "invalidsshkey"; }
```
[the remark this answers is not in the document]
Comment (faust64) [reply htdocs/changesshkey.php:45]: ok

File: tests/CheckSshkeyTest.php (hunk 5 above)
Comment (faust64) [reply tests/CheckSshkeyTest.php:18]: ok

File: conf/config.inc.php (hunk 6 above)
Comment (faust64) [reply conf/config.inc.php]: ok

File: lib/functions.inc.php [hunk 8]
```diff
@@ -519,6 +519,39 @@ function change_password( $ldap, $dn, $password, $ad_mode, $ad_options, $samba_m
...
+    }
+
+    $key = $key_parts[1];
+    $key_base64_decoded = base64_decode($key, true);
+    if ($key_base64_decoded == FALSE) {
+	return false;
+    }
+
+    $ealg = base64_encode($algorithm . " AAAA");
+    $check = preg_replace('/[\x00-\x1F\x7F]/u', '', base64_decode(substr($key, 0, strlen($ealg))));
```
[the remark this answers is not in the document]
Comment (faust64) [reply lib/functions.inc.php:547]: If you can show me a case where this wouldn't work, I'ld be glad to fix it.
I'ld rather not pull third-party libraries, nor run system commands with arbitrary user input.

File: htdocs/changesshkey.php (hunk 7 above)
[the remark this answers is not in the document]
Comment (plewin) [reply htdocs/changesshkey.php:45]: I meant invalidsshkey not sshkeyrequired my bad, thanks for noticing it.
I guess sshkeyrequired is danger color is also fine.

[the remark this answers is not in the document]
Comment (faust64) [reply htdocs/changesshkey.php:45]: I reverted the invalidsshkey one, which was already marked as a warning. Only keeping the sshkeyrequired in danger.

**Review by plewin — COMMENTED** [review]
Looks good to me, my only issue is the whitespace inconsistencies

**Review by plewin — APPROVED** [review]

**Comment by coudot** [comment]
We should indeed use spaces and not tabs.

In vi:
```:set tabstop=4 shiftwidth=4 expandtab

**Comment by coudot** [comment]
Tested on my side, works well.

Thanks @faust64

**Pull request merged**